minari/utils/data_collector.py

```diff
@@ -0,0 +1,573 @@
+from __future__ import annotations
+
+import os
+import shutil
+import tempfile
+from typing import Any, Dict, Optional, Type, Union
+
+import gymnasium as gym
+import h5py
+import numpy as np
+from gymnasium import spaces
+
+
+STEP_DATA_KEYS = {
+    "actions",
+    "observations",
+    "rewards",
+    "truncations",
+    "terminations",
+}
+
+
+class EpisodeMetadataCallback:
+    """Callback to full episode after saving to hdf5 file as a group.
+
+    This callback can be overridden to add extra metada attributes or statistics to
+    each HDF5 episode group in the Minari dataset. The custom callback can then be
+    passed to the DataCollectorV0 wrapper in the `episode_metadata_callback` argument.
+
+    TODO: add more default statistics to episode datasets
+    """
+
+    def __call__(self, eps_group: h5py.Group):
+        """Callback method.
+
+        Override this method to add custom attribute metadata to the episode group.
+
+        Args:
+            eps_group (h5py.Group): the HDF5 group that contains an episode's datasets
+        """
+        eps_group["rewards"].attrs["sum"] = np.sum(eps_group["rewards"])
+        eps_group["rewards"].attrs["mean"] = np.mean(eps_group["rewards"])
+        eps_group["rewards"].attrs["std"] = np.std(eps_group["rewards"])
+        eps_group["rewards"].attrs["max"] = np.max(eps_group["rewards"])
+        eps_group["rewards"].attrs["min"] = np.min(eps_group["rewards"])
+
+        eps_group.attrs["total_steps"] = eps_group["actions"].shape[0]
+
+
+class StepDataCallback:
+    """Callback to create step data dictionary from the return data of each Gymnasium environment step.
+
+    The current callback automatically detects observation/action spaces that need
+    to be flatten before saving to HDF5 file (currently only supports Dict or Tuple
+    Gymnasium spaces. Text, Sequence, and Graph are currently not compatible with
+    Minari).
+
+    This callback can be overridden to add extra environment information in each step or
+    edit the observation, action, reward, termination, truncation, or info returns.
+    """
+
+    def __init__(self, env: gym.Env):
+        self.env = env
+
+        def check_flatten_space(space: gym.spaces.Space):
```

Comment:
Can you use `space. is_np_flattenable `? https://gymnasium.farama.org/api/spaces/#gymnasium.spaces.space.Space.is_np_flattenable

Reply:
I can't because I need to check specifically for `Dict` and `Tuple` spaces. The rest of the spaces shouldn't be flattened or are not yet compatible with Minari